src/openforms/formio/tests/validation/test_file.py

```diff
@@ -0,0 +1,466 @@
+
+TEST_FILES_DIR = Path(__file__).parents[1] / "files"
+
+
+class _SubmittedFileDataFactory(factory.DictFactory):
+    url = factory.SelfAttribute("..url")
+    form = ""
+    name = factory.SelfAttribute("..name")
+    size = factory.SelfAttribute("..size")
+    baseUrl = factory.LazyAttribute(lambda d: furl(d.factory_parent.url).origin)
+    project = ""
+
+
+class SubmittedFileFactory(factory.DictFactory):
+    name = factory.Sequence(lambda n: f"foo-{n}.bin")
+    originalName = factory.LazyAttribute(lambda obj: obj.name)
+    size = factory.Faker("pyint", min_value=1)
+    storage = "url"
+    type = "application/foo"
+    url = "http://localhost/api/v2/submissions/files/123"
+    data = factory.SubFactory(_SubmittedFileDataFactory)
+
+    @factory.post_generation
+    def with_temporary_upload(obj, create, extracted, **kwargs):
+        if extracted:
+            # The temporary upload was explicitly provided, read the necessary values from there
+            temporary_upload = extracted
+            obj["originalName"] = obj["data"]["name"] = temporary_upload.file_name
+            obj["size"] = obj["data"]["size"] = temporary_upload.file_size
+        else:
+            create_temporary_upload = (
+                TemporaryFileUploadFactory.create
+                if create
+                else TemporaryFileUploadFactory.build
+            )
+            temporary_upload = create_temporary_upload(
+                file_name=obj["name"], file_size=obj["size"]
+            )
+        new_url = f"http://localhost/api/v2/submissions/files/{temporary_upload.uuid}"
+        obj["url"] = obj["data"]["url"] = new_url
+
+
+DEFAULT_FILE_COMPONENT: FileComponent = {
+    "type": "file",
+    "key": "foo",
+    "label": "Test",
+    "storage": "url",
+    "url": "",
+    "useConfigFiletypes": False,
+    "filePattern": "",
+    "file": {"allowedTypesLabels": []},
+}
+
+
+class FileValidationMaxFilesAndRequiredTests(TestCase):
+    """Tests related to ``validate.required`` and ``maxNumberOfFiles``."""
+
+    def test_file_not_required(self):
+        component: FileComponent = {
+            "type": "file",
+            "key": "foo",
+            "label": "Test",
+            "storage": "url",
+            "url": "",
+            "useConfigFiletypes": False,
+            "filePattern": "",
+            "file": {"allowedTypesLabels": []},
+        }
+
+        is_valid, _ = validate_formio_data(component, {})
+
+        self.assertTrue(is_valid)
+
+    def test_file_required(self):
+        component: FileComponent = {
+            "type": "file",
+            "key": "foo",
+            "label": "Test",
+            "storage": "url",
+            "url": "",
+            "useConfigFiletypes": False,
+            "filePattern": "",
+            "file": {"allowedTypesLabels": []},
+            "validate": {"required": True},
+        }
+
+        is_valid, errors = validate_formio_data(component, {})
```

Comment:
we definitely need to see more cases that represent formio's real behaviour. I suspect it will send a value `[]` or maybe even `[null]` if left blank in the UI